src/aspire/basis/fspca.py

```diff
@@ -0,0 +1,550 @@
+            basis_inds = np.arange(eigval_index, eigval_index + len(eigvals_k))
+
+            # Store the eigvals for this block, note this is a flat array.
+            self.eigvals[basis_inds] = eigvals_k
+
+            # Store the eigvecs, note this is a BlkDiagMatrix and is assigned incrementally.
+            self.eigvecs[angular_index] = eigvecs_k
+
+            # To compute new expansion coefficients using spca basis
+            #   we combine the basis coefs using the eigen decomposition.
+            # Note image stack slow moving axis, otherwise this is just a
+            #   block by block matrix multiply, with the einsum
+            #   indexing performing transposes for us.
+            self.spca_coef[:, basis_inds] = np.einsum(
+                "ji, kj -> ki", eigvecs_k, A[angular_index]
+            )
+
+            eigval_index += len(eigvals_k)
+
+        # Sanity check we have same dimension of eigvals and basis coefs.
+        if eigval_index != self.basis.count:
+            raise RuntimeError(
+                f"eigvals dimension {eigval_index} != basis coef count {self.basis.count}."
+            )
+
+        # Store a map of indices sorted by eigenvalue.
+        #  We don't resort then now because this would destroy the block diagonal structure.
+        #
+        # sorted_indices[i] is the ith most powerful eigendecomposition index
+        #
+        # We can pass a full or truncated slice of sorted_indices to any array indexed by
+        # the coefs.  This is used later for compression and index re-generation.
+        self.sorted_indices = np.argsort(-np.abs(self.eigvals))
+
+    def expand_from_image_basis(self, x):
+        """
+        Take an image in the standard coordinate basis and express as FSPCA coefs.
+
+        Note each FSPCA coef corresponds to a linear combination Fourier Bessel
+        basis vectors, described by an eigenvector in FSPCA.
+
+        :param x:  The Image instance representing a stack of images in the
+        standard 2D coordinate basis to be evaluated.
+        :return: Stack of coefs in the FSPCABasis.
+        """
+        fb_coefs = self.basis.evaluate_t(x)
+        return self.expand(fb_coefs)
+
+    def expand(self, x):
+        """
+        Take a Fourier-Bessel coefs and express as FSPCA coefs.
+
+        Note each FSPCA coef corresponds to a linear combination Fourier Bessel
+        basis vectors, described by an eigenvector in FSPCA.
+
+        :param x:  Coefs representing a stack in the
+        Fourier Bessel basis.
+        :return: Stack of coefs in the FSPCABasis.
+        """
+        # apply linear combination defined by FSPCA (eigvecs)
+        #  can try blk_diag here, but I think needs to be extended to non square...,
+        #  or masked.
+        # c_fspca = (self.eigvecs.apply(c_fb.T)).T
+        eigvecs = self.eigvecs
+        if isinstance(eigvecs, BlkDiagMatrix):
+            eigvecs = eigvecs.dense()
```

Comment:
I don't believe it will work when compressed... because eigvecs will almost certainly not be square block diagonal then.  In this way the original matlab was probably a little better (but of course it was bespoke for every problem).

I think we could fix this either by extending the block diagonal to non-square, or extending it with masking.  Both would achieve the same effect.  If you agree I will create an issue for this.

Reply:
#442